### ISSUE TITLE
Add option to send transactions private using Eden

### ISSUE THREAD

**Issue by hboon** [AlphaWallet/alpha-wallet-android#2318]
@oa-s would you help to elaborate? Note that checking for pending transaction status needs to be updated too.

**Comment by oa-s** [comment]
Its need to update code base with ability to use Eden private network. Checkout RPC url while sending a transaction to network.
The icons need for PR could find here: https://app.zeplin.io/project/5d0356be2f25b319dcc5944b/styleguide/components?coid=6197a1c20ba031a7c0eb8347
Note, in case if android doesn't support sending requests for `ethermine`, check RPC url for seach server:
`.main` - `https://rpc.ethermine.org`;
`other servers` - use default RPC url.
For .eden use the next private RPC urls:
`.main` - `https://api.edennetwork.io/v1/rpc`;
`ropsten` - `https://dev-api.edennetwork.io/v1/rpc`;
`other servers` - use default RPC url.

**Comment by hboon** [comment]
> Note, in case if android doesn't support sending requests for ethermine, check RPC url for seach server:

No, it doesn't yet. Might as well do it together.

@oa-s thanks. Help assign this to @JamesSmartCell when you are done. Maybe screenshots of where the changes are in the UI too?

**Comment by hboon** [comment]
@oa-s bump

**Comment by asif-finimble** [comment]
@JamesSmartCell can you please guide me on this?   
seems like edennetwork website is not accessible right now.  Also I dont have the access to the given zeplin project.

**Comment by JamesSmartCell** [comment]
@oa-s or @hboon can you show the current flow on iOS? Is there any special handling or is it just done as a new network as normal? 

If it's a normal network, I see RPC calls but no transaction query API routes. Do these correspond to existing networks? Or a single existing network eg Mainnet?

So, eg would I send the transaction via eden, get a tx hash and then scan for that on mainnet?

"Note, in case if android doesn't support sending requests for ethermine" what does this mean? Why would Android natively support any crypto transaction mechanism?

Sorry I don't really understand any of this PR or how it's supposed to work. Can you clarify with a quick run through for Asif and myself?

**Comment by hboon** [comment]
>@oa-s or @hboon can you show the current flow on iOS? Is there any special handling or is it just done as a new network as normal?

> If it's a normal network, I see RPC calls but no transaction query API routes. Do these correspond to existing networks? Or a single existing network eg Mainnet?

> So, eg would I send the transaction via eden, get a tx hash and then scan for that on mainnet?

Sure. This isn't the same as adding a chain that we normally do.

What Eden and Ethermine does is, they are miners and they offer an RPC endpoint for us to post transactions to. These transactions are only broadcast after they are mined, to prevent front-running (by others). Hence "private". They also offer some other functionality that involves staking, but we aren't using those.

So technically, if the user has to use Eden, everything stays the same, except when the user sends a transaction on Ethereum mainnet:

* the RPC endpoint to post the transaction should be Eden's (and not Infura)
* when checking the transaction status, the RPC endpoint should be Eden's too

This should be built, keeping in mind that we support more than one of this. In fact, we have Eden and Ethermine:

* Eden — Ethereum mainnet and Ethereum ropsten
* Ethermine — Ethereum mainnet

The 3 RPC endpoints are mentioned in the 2nd comment above.

**Comment by asif-finimble** [comment]
Thanks @hboon for the detailed explanation. I think I understand what needs to be done. How does the user select he wants to use Eden or ethermine? 
Maybe its in zeplin spec. I can't access that yet.

**Comment by hboon** [comment]
@asif-finimble Sorry, missed this. Like this:

<img width="200" alt="Screenshot 2022-03-02 at 10 09 21 PM" src="https://user-images.githubusercontent.com/56189/156377719-41d8188d-544f-4b04-bf63-dcfdbfeb05fc.png">

⬇️

<img width="200" alt="Screenshot 2022-03-02 at 10 09 26 PM" src="https://user-images.githubusercontent.com/56189/156377745-f864a273-da87-4256-a856-ba6ae1e465fd.png">

⬇️

<img width="200" alt="Screenshot 2022-03-02 at 10 09 27 PM" src="https://user-images.githubusercontent.com/56189/156377750-ba887104-10e2-4570-b1a6-ea59fda408e7.png">

Enabled (will look a bit different for Android):

<img width="200" alt="Screenshot 2022-03-02 at 10 11 44 PM" src="https://user-images.githubusercontent.com/56189/156377942-05307ee2-b3f1-4182-bdfc-8167ca27b94f.png">

**Comment by asif-finimble** [comment]
@hboon how can the user disable private transactions? Does the selection state toggles on clicking the selected item? @JamesSmartCell what view do you think we should use here?

**Comment by hboon** [comment]
@asif-finimble Yes, it should toggle to off if the user taps on a selected item.

I don't know how to present it though.

**Comment by asif-finimble** [comment]
> @asif-finimble Yes, it should toggle to off if the user taps on a selected item.
> 
> I don't know how to present it though.

OK. Thanks. Got it

**Comment by asif-finimble** [comment]
@hboon is this setting common for all wallets(if the user has multiple wallets) or we need to store the setting separately for each wallet?

**Comment by hboon** [comment]
@asif-finimble Up to @JamesSmartCell  , we just make it global for all users on iOS.

Eden icon here: https://github.com/AlphaWallet/alpha-wallet-ios/issues/3445
Ethermine's: https://github.com/AlphaWallet/alpha-wallet-ios/issues/3270#issuecomment-953592854

For iOS, we just use the selected option's (Eden or Ethermine) icon in Settings > Advanced. If none is chose, use Ethermine's

**Comment by asif-finimble** [comment]
Thanks @hboon.  
As for the icons, the ticket contains link to zeplin which I don't have access to.

**Comment by asif-finimble** [comment]
@JamesSmartCell As per my understanding, we need to add new chain IDs and corresponding ```NetworkInfo``` objects for Eden and Ethermine in the ```networkMap```.   And we will check at the time of sending transactions, if private transactions are enabled, then we will use the selected private network's rpc for sending on mainnet and ropsten. Am I Right? or can you suggest appropriate flow.

**Comment by hboon** [comment]
> Thanks @hboon.
> As for the icons, the ticket contains link to zeplin which I don't have access to.

How about this? https://app.zeplin.io/project/5d088205bff2d15de6a4397b/styleguide/components?coid=6197a1c20ba031a7c0eb8347 (the 1st 2 icons). You can get here from the Android project in Zeppelin.

**Comment by asif-finimble** [comment]
> > Thanks @hboon.
> > As for the icons, the ticket contains link to zeplin which I don't have access to.
> 
> How about this? https://app.zeplin.io/project/5d088205bff2d15de6a4397b/styleguide/components?coid=6197a1c20ba031a7c0eb8347 (the 1st 2 icons). You can get here from the Android project in Zeppelin.

Thanks. This will do.

**Comment by asif-finimble** [comment]
> @JamesSmartCell As per my understanding, we need to add new chain IDs and corresponding `NetworkInfo` objects for Eden and Ethermine in the `networkMap`. And we will check at the time of sending transactions, if private transactions are enabled, then we will use the selected private network's rpc for sending on mainnet and ropsten. Am I Right? or can you suggest appropriate flow.

I get it that the chainId will remain 1 - MAINNET_ID.  But where do we specify to use the private rpc.?
- One way i could think is, ```NetworkInfo``` will contain the private networks rpc, and ```NetworkInfo``` will return private or public rpc url based on whether private txn is active. We can update the ```getWeb3jService``` method for deciding which rpc url to use.     
- we'll also need to update ```Transaction``` class to record whether this was a private or public transaction so when we show it from cache, we can call appropriate rpc for status.    
Let me know what do you think.

**Comment by JamesSmartCell** [comment]
@asif-finimble I refactored the code to make this slightly simpler - although you will now have to accomodate EIP1559 and Legacy transactions. Ping me for more info if required, but I suspect the method you'll be modifying is this code:

```
EthSendTransaction raw = web3j
							.ethSendRawTransaction(Numeric.toHexString(signedMessage.signature))
							.send();
```

You probably want to abstract the 3 uses of this out into a separate function which can switch between sending to Eden/Ethermine or default chain.
